### PR TITLE
[Function Builders] Teach diagnostics about function builder body result types...

### PR DESCRIPTION
**Pull request by hborla**
...in order to properly diagnose requirement failures that aren't anchored at an expression.

* Add a new locator path element `FunctionBuilderBodyResult`
* When adding the constraint to match the function builder transformed type with the body result type, anchor at the function decl and add the `FunctionBuilderBodyResult` path element.
* When computing the affected decl for a `RequirementFailure` that has a locator pointing to a `FunctionBuilderBodyResult`, get the affected decl from the function result type
* Remove an unused parameter to `matchFunctionBuilder`

Resolves: rdar://problem/65683605

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintLocator.cpp [hunk 1]
```diff
@@ -249,6 +250,11 @@ bool ConstraintLocator::isForOptionalTry() const {
   return directlyAt<OptionalTryExpr>();
 }
 
+bool ConstraintLocator::isForFunctionBuilderBodyResult() const {
+  auto elt = getFirstElementAs<LocatorPathElt::FunctionBuilderBodyResult>();
+  return elt.hasValue();
```
Comment (xedin) [inline lib/Sema/ConstraintLocator.cpp:255]: Nit: `ConstraintLocator` has `isLastElement` helper which could be used here.

**Review by xedin — APPROVED** [review]
Looks great!

**Comment by hborla** [comment]
@swift-ci please smoke test OS X Platform

File: lib/Sema/ConstraintLocator.cpp (hunk 1 above)
Comment (hborla) [reply lib/Sema/ConstraintLocator.cpp:255]: There's no equivalent `isFirstElement` - should I add one?

Comment (xedin) [reply lib/Sema/ConstraintLocator.cpp:255]: Yeah, I think that would be useful!

Comment (hborla) [reply lib/Sema/ConstraintLocator.cpp]: ✅

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test Linux Platform

**Pull request merged**